File: 0048-rotate-image/0048-rotate-image.py
class Solution:
    def reverse(self, matrix, index):
        i= 0 
        j = len(matrix[index])-1
        while i<=j:
            matrix[index][j],  matrix[index][i] = matrix[index][i], matrix[index][j]
            i+=1
            j-=1
        
    def rotate(self, matrix: List[List[int]]) -> None:
        """
        Do not return anything, modify matrix in-place instead.
        """
        n = len(matrix)
        for i in range(n):
            for j in range(i+1, n):
                matrix[i][j], matrix[j][i] = matrix[j][i], matrix[i][j]
        print(matrix)
        for i in range(n):
            self.reverse(matrix, i)
        return matrix 

File: 0048-rotate-image/0048-rotate-image.java
class Solution {
    public static int[][] reverse(int matrix[][], int n){
        int arr[] = matrix[n];
        int left = 0;
        int right = arr.length-1;
        while(left<right){
            int temp = arr[left];
            arr[left] = arr[right];
            arr[right]=temp;
            left+=1;
            right-=1;
        } 
        matrix[n] = arr;
        return matrix;
    }
    public void rotate(int[][] matrix) {
        //transpose
        for(int i=0;i<matrix.length;i++){
            for(int j=i;j<matrix[0].length;j++){
                if(i==j) continue;
                int temp = matrix[i][j];
                matrix[i][j]=matrix[j][i];
                matrix[j][i] = temp;
            }
        }
        for(int i=0;i<matrix.length;i++){
            matrix = reverse(matrix, i);
        }
    }
}